src/Components/Server/src/ComponentHub.cs

```diff
@@ -389,6 +395,39 @@ static bool RootComponentIsEmpty(string rootComponents) =>
             string.IsNullOrEmpty(rootComponents) || rootComponents == "[]";
     }
 
+    // Client initiated pauses work as follows:
+    // * The client calls PauseCircuit, we dissasociate the circuit from the connection.
+    // * We trigger the circuit pause to collect the current root components and dispose the current circuit.
+    // * We push the current root components and application state to the client.
+    //   * If that succeeds, the client receives the state and we are done.
+    //   * If that fails, we will fall back to the server-side cache storage.
+    // * The client will disconnect after receiving the state or after a 30s timeout.
+    //   * From that point on, it can choose to resume the circuit by calling ResumeCircuit with or without the state
```

Comment:
What happens if client calls the `resume()` in the middle of pending pause->push to client workflow?

Reply:
We push to the client when the client initiates the pause explicitly, so at that point the client must have called `pause()` which `resume()` is aware of and will await until `pause()` completes before trying the `resume()` operation.